### PR TITLE
Added support for custom parameters. Added test class for coverage

### PR DESCRIPTION
**Pull request by sander-adhese**

### CODE REVIEW

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 1]
```diff
@@ -101,6 +117,24 @@ public Builder withCookieMode(CookieMode cookieMode) {
...
+                throw new IllegalArgumentException("To add a valid custom parameter, your key must be two chars long and have at least one value.");
+            }
+
+            if (this.options.customParameters.containsKey(key)) {
+                this.options.customParameters.get(key).addAll(values);
+            } else {
+                this.options.customParameters.put(key, new HashSet<>(values));
+            }
+
+            return this;
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java:148]: It's the responsibility of the caller to html escape the values ?

Comment (sander-adhese) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java:148]: Perhaps not. I'll add that, and the 'illegal' chars validation

Comment (paulvv) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java:148]: but what about the other target values? were they escaped?

Comment (sander-adhese) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java:148]: no they are not. So maybe it is their responsibility

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 2]
```diff
@@ -59,6 +64,15 @@ public String getAsURL() {
         }
 
         builder.append(String.format("/%s%s", AdheseParameter.COOKIE_MODE.getKey(), cookieMode.getValue()));
+        TreeMap<String, Set<String>> sorted = new TreeMap<>(customParameters);
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: Isn't this overkill? why not just sort the keys?

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 3]
```diff
@@ -59,6 +64,15 @@ public String getAsURL() {
...
+        for (Map.Entry<String, Set<String>> customParameter: sorted.entrySet()) {
+            StringBuilder values = new StringBuilder();
+            for (String value: customParameter.getValue()) {
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: these also need sorting to have deterministic tests

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 4]
```diff
@@ -101,6 +115,42 @@ public Builder withCookieMode(CookieMode cookieMode) {
...
+        public Builder addCustomParametersRaw(Map<String, ? extends Collection<String>> map) {
+            for (Map.Entry<String, ? extends Collection<String>> entry: map.entrySet()) {
+                this.addCustomParameterRaw(entry.getKey(), entry.getValue());
+            }
+
+            return this;
+        }
+
+        public Builder removeCustomParameters() {
+            this.options.customParameters = Collections.emptyMap();
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: why not map.clear ?

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 5]
```diff
@@ -101,6 +115,42 @@ public Builder withCookieMode(CookieMode cookieMode) {
...
+            return this.addCustomParameterRaw(key, Arrays.asList(value));
+        }
+
+        public Builder addCustomParameterRaw(String key, Collection<String> values) {
+            if (key == null || values == null || key.length() != 2 || values.isEmpty()) {
+                throw new IllegalArgumentException("To add a valid custom parameter, your key must be two chars long and have at least one value.");
+            }
+
+            if (this.options.customParameters.containsKey(key)) {
+                this.options.customParameters.get(key).addAll(values);
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: do we need all the this. prefixing ?

File: sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        builder.addCustomParameterRaw("zz", Arrays.asList("abc"));
+        builder.addCustomParameterRaw("aa", Arrays.asList("abc"));
+        builder.addCustomParameterRaw("oo", Arrays.asList("abc"));
+        assertEquals("/slloc1-slot1/tlnone/aaabc/ooabc/zzabc", builder.build().getAsURL());
+    }
+
+    @Test( expected = IllegalArgumentException.class)
+    public void custom_parameter_validate_key() {
+        AdheseOptions.Builder builder = getBaseBuilder();
+        builder.addCustomParameterRaw("xab", Arrays.asList("abc", "def"));
```
Comment (paulvv) [inline sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java:81]: we don't have List.of ?

**Review by paulvv — CHANGES_REQUESTED** [review]

File: sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java (hunk 6 above)
Comment (sander-adhese) [reply sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java:81]: we don't

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java (hunk 5 above)
Comment (sander-adhese) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: we don't

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java (hunk 4 above)
Comment (sander-adhese) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: em yes that's better

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java (hunk 3 above)
Comment (sander-adhese) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: how do you mean? what would you test that eg `custom_parameter_case_list_multiple` doesn't cover?

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 7]
```diff
@@ -59,6 +64,15 @@ public String getAsURL() {
         }
 
         builder.append(String.format("/%s%s", AdheseParameter.COOKIE_MODE.getKey(), cookieMode.getValue()));
+        TreeSet<String> sortedKeys = new TreeSet<>(customParameters.keySet());
+        for (String key: sortedKeys) {
+            StringBuilder values = new StringBuilder();
+            for (String value: customParameters.get(key)) {
+                values.append(String.format("%s%s", value, ";"));
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: code like this makes want to abuse you with Tropic Thunder quotes. don't add the first semi-colon, so you don't have to delete it later...

File: sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java (hunk 6 above)
Comment (paulvv) [reply sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java:81]: :-(

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java (hunk 7 above)
Comment (sander-adhese) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: I assume you mean the last one? I'll change the loop

Comment (paulvv) [reply sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: thumbs up icon

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 8]
```diff
@@ -59,6 +64,20 @@ public String getAsURL() {
         }
 
         builder.append(String.format("/%s%s", AdheseParameter.COOKIE_MODE.getKey(), cookieMode.getValue()));
+        TreeSet<String> sortedKeys = new TreeSet<>(customParameters.keySet());
+        for (String key: sortedKeys) {
+            StringBuilder values = new StringBuilder();
+            List<String> valueList = new ArrayList(customParameters.get(key));
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java:70]: these also need sorting

File: sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        AdheseOptions.Builder builder = getBaseBuilder();
+        builder.addCustomParameterRaw("xa", "abc");
+        assertEquals("/slloc1-slot1/tlnone/xaabc", builder.build().getAsURL());
+    }
+
+    @Test
+    public void custom_parameter_case_map() {
+        AdheseOptions.Builder builder = getBaseBuilder();
+        Map map = new HashMap<String, List<String>>();
+        map.put("xa", Arrays.asList("abc", "def", "ghi"));
```
Comment (paulvv) [inline sdk/src/test/java/com/endare/adhese/sdk/AdheseOptionsTest.java]: reorder these to see the sorting of values work

File: sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java [hunk 10]
```diff
@@ -59,6 +64,20 @@ public String getAsURL() {
         }
 
         builder.append(String.format("/%s%s", AdheseParameter.COOKIE_MODE.getKey(), cookieMode.getValue()));
+        TreeSet<String> sortedKeys = new TreeSet<>(customParameters.keySet());
+        for (String key: sortedKeys) {
+            StringBuilder values = new StringBuilder();
+            List<String> valueList = new ArrayList(new TreeSet(customParameters.get(key)));
```
Comment (paulvv) [inline sdk/src/main/java/com/endare/adhese/sdk/AdheseOptions.java]: using Collections.sort we can have one Object instantiation less

**Comment by paulvv** [comment]
ok

**Comment by sander-adhese** [comment]
So what now? Can I just merge this? Do I need to bump some version somewhere?

**Comment by paulvv** [comment]
Tim made some changes previously, so check with him

**Pull request merged**